netty/src/main/java/io/grpc/netty/NettyServerHandler.java

```diff
@@ -188,16 +188,17 @@ private void onHeadersRead(ChannelHandlerContext ctx, int streamId, Http2Headers
       // method.
       Http2Stream http2Stream = requireHttp2Stream(streamId);
 
-      NettyServerStream stream = new NettyServerStream(ctx.channel(), http2Stream, this,
-              maxMessageSize);
+      NettyServerStream.TransportState state =
+          new NettyServerStream.TransportState(this, http2Stream, maxMessageSize);
+      NettyServerStream stream = new NettyServerStream(ctx.channel(), state);
 
       Metadata metadata = Utils.convertHeaders(headers);
-      stream.inboundHeadersReceived(metadata);
 
       ServerStreamListener listener =
           transportListener.streamCreated(stream, method, metadata);
-      stream.setListener(listener);
-      http2Stream.setProperty(streamKey, stream);
+      // TODO(ejona): this is racy since stream could have been used before getting here.
+      state.setListener(listener);
```

Comment:
Is this safe to be left as a TODO? That could realistically lead to NPE with `listener.(halfClosed|closed)()`?

We should maybe initialize the TransportState with a listener that remembers any callbacks and replay's them once `setListener` is called? Or do that in a follow up PR?